tests/test_forward_ref.py

```diff
@@ -840,3 +859,95 @@ class Foo(BaseModel):
     bar_model = module.nested()
     assert bar_model.__pydantic_model_complete__ is True
     assert bar_model(b={'a': 1}).dict() == {'b': {'a': 1}}
+
+
+def pytest_raises_undefined_types_warning(defining_class_name, missing_type_name):
+    """Returns a pytest.raises context manager that checks for the correct undefined types warning message.
+    usage: `with pytest_raises_undefined_types_warning(class_name='Foobar', missing_class_name='UndefinedType'):`
+    """
+
+    return pytest.raises(
+        UserWarning,
+        match=re.escape(
+            # split this f-string into two lines to pass linting (line too long)
+            f'`{defining_class_name}` is not fully defined, you should define `{missing_type_name}`, '
+            f'then call `{defining_class_name}.model_rebuild()`',
+        ),
+    )
+
+
+def test_undefined_type_raises_warning(create_module):
+    """Test undefined_types_warning is raised when a model declares a variable of type that is missing."""
+
+    with pytest_raises_undefined_types_warning(defining_class_name='Foobar', missing_type_name='UndefinedType'):
+        create_module(
+            # language=Python
+            """
+from __future__ import annotations
+from pydantic import BaseModel
+
+class Foobar(BaseModel):
+    a: UndefinedType
+"""
+        )
+
+
+def test_undefined_type_raises_warning_if_forward_ref(create_module):
+    """Test undefined_types_warning is raised when a model declares a variable of type that is a ForwardRef."""
+
+    with pytest_raises_undefined_types_warning(defining_class_name='Foobar', missing_type_name='UndefinedType'):
+
+        @create_module
+        def module():
+            from typing import ForwardRef
+
+            from pydantic import BaseModel
+
+            UndefinedType = ForwardRef('UndefinedType')
+
+            class Foobar(BaseModel):
+                a: UndefinedType
+
+
+def test_disable_undefined_types_warning(create_module):
+    """Test suppressing undefined_types_warning via config \
+        when a model declares a variable of type that is missing."""
+
+    module = create_module(
+        # language=Python
+        """
+from __future__ import annotations
+from pydantic import BaseModel
+
+class Foobar(BaseModel):
+    a: UndefinedType
+    class Config:
+        undefined_types_warning = False
+"""
+    )
+    # Since test is about suppressing the undefined_types_warning,
+    # we should also verify that our model was at least partially built
+    assert module.Foobar.__pydantic_model_complete__ is False
```

Comment:
Nice test. Looks very similar to the test `test_undefined_type_raises_warning()`. There are 2 general ways to make life simpler here:
1. Describe the relationship of these 2 tests in their docstrings. (Human talking to a human solution)
2. Parameterize these into a single test. In this case, you get tests which is slightly DRY-er, but slightly more more complex.

I personally am in favor of the 2nd option. Btw, feel free to delegate to a follow-up PR which I would be happy to contribute to – OR, I can contribute to this PR if you want support with this. (OR, choose option 1.) _OR..._ something else?

Reply:
@ybressler I see what you mean, yet I'm not quite sure how to go about parameterizing a test that use `create_module`. Any code defined outside the `def module()` function is easily accessible inside the `module` that is the code under test.  Also, for tests where the `module` is calling `from __future__ import annotations` we need to `create_module()` from a string `# language=Python` otherwise it seems the tests don't run correctly (I'm not entirely sure why, but generally `from __future__ import annotations` has the be the first line of code in a file).  So, for the sake of symmetry and readability, I'm inclined the just sprinkle in some additional comments (1st option you recommended).  All that said, I'm very curious what changes you'd propose (since I'd probably learn a lot and you're probably able to see further around the obstacles than I can) -- I would probably defer that to a follow-up PR, once the smoke settles around the questions of ForwardRef handling (in Python < 3.10 vs. >=3.10).

Reply:
Good approach. Once comments are added, let's close this thread.

Reply:
comments added. I'll close this thread and we can review the changes in that other thread.